lib/ecto/embedded.ex

```diff
@@ -64,39 +63,108 @@ defmodule Ecto.Embedded do
   @doc """
   Wraps embedded models in changesets.
   """
-  def change(%Embedded{cardinality: :one}, value) do
-    Ecto.Changeset.change(value)
+  def change(%Embedded{cardinality: :one}, value, current) do
+    do_change(value, current)
   end
 
-  def change(%Embedded{cardinality: :many, container: :array}, value) do
-    Enum.map(value, &Ecto.Changeset.change/1)
+  def change(%Embedded{cardinality: :many, container: :array, embed: mod},
+             value, current) when is_list(value) do
+    {pk, _} = primary_key(mod)
+    current = process_current(current, pk)
+    map_changes(value, pk, &do_change/2, current, [], true) |> elem(1)
+  end
+
+  defp do_change(value, nil) do
+    %{Changeset.change(value) | action: :insert}
```

Comment:
Because we know the action is insert, we should load all the fields in the model as changes. This have exactly the same logic as in Repo.insert, so we can discuss if we should perform this change here or in Repo.insert (likely Repo.insert).

The same logic applies for update.